Not able to link with pyo3 in debug mode on Windows

### Bug Description

In my debug builds, for years, I used to compile my third-parties from the source code in debug mode.

For example, I compile python 3.10.8 in debug mode and it produces:
python_d.exe
python3_d.dll
python310_d.dll
python3_d.lib
python310_d.lib

On my build chain, I would like to add the python [cryptography](https://github.com/pyca/cryptography) third-party. When I compile it from the source code in debug mode, I have the following error:
```
python_d -m build --wheel --no-isolation -C--build-option="build --debug"
...
     Running `rustc --crate-name cryptography_rust --edition=2018 src\lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type cdylib --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 --crate-type cdylib --cfg "feature=\"default\"" --cfg "feature=\"extension-module\"" -C metadata=d36a220355622476 --out-dir C:\python-modules\cryptography\src\rust\target\debug\deps -C incremental=C:\python-modules\cryptography\src\rust\target\debug\incremental -L dependency=C:\python-modules\cryptography\src\rust\target\debug\deps --extern asn1=C:\python-modules\cryptography\src\rust\target\debug\deps\libasn1-89a4a2858d227583.rlib --extern chrono=C:\python-modules\cryptography\src\rust\target\debug\deps\libchrono-9c0c4c35fe6d5120.rlib --extern once_cell=C:\python-modules\cryptography\src\rust\target\debug\deps\libonce_cell-d359c42b90186127.rlib --extern ouroboros=C:\python-modules\cryptography\src\rust\target\debug\deps\libouroboros-818767ba07ca8884.rlib --extern pem=C:\python-modules\cryptography\src\rust\target\debug\deps\libpem-9d903dcf6b15ed52.rlib --extern pyo3=C:\python-modules\cryptography\src\rust\target\debug\deps\libpyo3-78c5487a10cfe0e9.rlib -L native=C:\usrlib\python\libs`
...
   = note: LINK : fatal error LNK1181: cannot open input file 'python3.lib'
```

This error is coming from the following line:
https://github.com/PyO3/pyo3/blob/56f11c252dd3382bee0611b085dabc437996d2c7/pyo3-build-config/src/impl_.rs#L1536

If I change `WINDOWS_ABI3_LIB_NAME` to `"python310_d"`, it is linking fine and the cryptography third-party is working.

However, if I change `WINDOWS_ABI3_LIB_NAME` to `"python3_d"`, it is linking fine but I have an error on the runtime:
`ImportError: Module use of python3_d.dll conflicts with this version of Python.`

This error is coming the following line in [cpython](https://github.com/python/cpython) repo:
https://github.com/python/cpython/blob/5f4ae86a639fb84260d622e31468da21dc468265/Python/dynload_win.c#L309-L330

The variable `buffer` is equals to `python310_d.dll` and the variable `import_python` is equals to `python3_d.dll`.

Could the linking of PyO3 in debug mode on Windows be fixed?


### Steps to Reproduce

1. Compile python in debug mode
2. Compile python cryptography in debug mode

### Backtrace

```shell
Here is the backtrace of the runtime error when I change WINDOWS_ABI3_LIB_NAME to "python3_d":
python_d.exe -vvvvvvvvv -c "from cryptography.fernet import Fernet"
...
# trying C:\usrlib\python\Lib\site-packages\cryptography\hazmat\bindings\_rust_d.cp310-win_amd64.pyd
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\usrlib\python\Lib\site-packages\cryptography\fernet.py", line 14, in <module>
    from cryptography.hazmat.primitives import hashes, padding
  File "<frozen importlib._bootstrap>", line 1078, in _handle_fromlist
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\usrlib\python\Lib\site-packages\cryptography\hazmat\primitives\padding.py", line 11, in <module>
    from cryptography.hazmat.bindings._rust import (
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 674, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1176, in create_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
ImportError: Module use of python3_d.dll conflicts with this version of Python.
```


### Your operating system and version

Windows 10 21H2 (OS Build 19044.2130)

### Your Python version (`python --version`)

Python 3.10.8 (in debug mode)

### Your Rust version (`rustc --version`)

rustc 1.65.0 (897e37553 2022-11-02)

### Your PyO3 version

0.15.2

### How did you install python? Did you use a virtualenv?

Compilation from the source code in debug mode.

### Additional Info

_No response_

--- a/newsfragments/2937.fixed.md
+++ b/newsfragments/2937.fixed.md
@@ -1,0 +1,1 @@
+Link against `pythonXY_d.dll` for debug Python builds on Windows.

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -230,6 +230,7 @@ print_if_set("base_prefix", base_prefix)
 print("executable", sys.executable)
 print("calcsize_pointer", struct.calcsize("P"))
 print("mingw", get_platform().startswith("mingw"))
+print("ext_suffix", get_config_var("EXT_SUFFIX"))
 "#;
         let output = run_python_script(interpreter.as_ref(), SCRIPT)?;
         let map: HashMap<String, String> = parse_script_output(&output);
@@ -261,6 +262,10 @@ print("mingw", get_platform().startswith("mingw"))
                 implementation,
                 abi3,
                 map["mingw"].as_str() == "True",
+                // This is the best heuristic currently available to detect debug build
+                // on Windows from sysconfig - e.g. ext_suffix may be
+                // `_d.cp312-win_amd64.pyd` for 3.12 debug build
+                map["ext_suffix"].starts_with("_d."),
             )
         } else {
             default_lib_name_unix(
@@ -1418,6 +1423,7 @@ fn default_abi3_config(host: &Triple, version: PythonVersion) -> InterpreterConf
             implementation,
             abi3,
             false,
+            false,
         ))
     } else {
         None
@@ -1493,6 +1499,7 @@ fn default_lib_name_for_target(
             implementation,
             abi3,
             false,
+            false,
         ))
     } else if is_linking_libpython_for_target(target) {
         Some(default_lib_name_unix(version, implementation, None))
@@ -1506,8 +1513,13 @@ fn default_lib_name_windows(
     implementation: PythonImplementation,
     abi3: bool,
     mingw: bool,
+    debug: bool,
 ) -> String {
-    if abi3 && !implementation.is_pypy() {
+    if debug {
+        // CPython bug: linking against python3_d.dll raises error
+        // https://github.com/python/cpython/issues/101614
+        format!("python{}{}_d", version.major, version.minor)
+    } else if abi3 && !implementation.is_pypy() {
         WINDOWS_ABI3_LIB_NAME.to_owned()
     } else if mingw {
         // https://packages.msys2.org/base/mingw-w64-python
@@ -2244,7 +2256,8 @@ mod tests {
                 PythonVersion { major: 3, minor: 7 },
                 CPython,
                 false,
-                false
+                false,
+                false,
             ),
             "python37",
         );
@@ -2253,7 +2266,8 @@ mod tests {
                 PythonVersion { major: 3, minor: 7 },
                 CPython,
                 true,
-                false
+                false,
+                false,
             ),
             "python3",
         );
@@ -2262,7 +2276,8 @@ mod tests {
                 PythonVersion { major: 3, minor: 7 },
                 CPython,
                 false,
-                true
+                true,
+                false,
             ),
             "python3.7",
         );
@@ -2271,7 +2286,8 @@ mod tests {
                 PythonVersion { major: 3, minor: 7 },
                 CPython,
                 true,
-                true
+                true,
+                false,
             ),
             "python3",
         );
@@ -2280,9 +2296,32 @@ mod tests {
                 PythonVersion { major: 3, minor: 7 },
                 PyPy,
                 true,
-                false
+                false,
+                false,
             ),
             "python37",
+        );
+        assert_eq!(
+            super::default_lib_name_windows(
+                PythonVersion { major: 3, minor: 7 },
+                CPython,
+                false,
+                false,
+                true,
+            ),
+            "python37_d",
+        );
+        // abi3 debug builds on windows use version-specific lib
+        // to workaround https://github.com/python/cpython/issues/101614
+        assert_eq!(
+            super::default_lib_name_windows(
+                PythonVersion { major: 3, minor: 7 },
+                CPython,
+                true,
+                false,
+                true,
+            ),
+            "python37_d",
         );
     }
 

